specs/logging_implementation.adoc

```diff
@@ -0,0 +1,274 @@
+
+== Other end user impact
+
+None
+
+== Performance Impact
+
+None
+
+== Other deployer impact
+
+None
+
+== Developer impact
+
+None
+
+== Implementation
+
+* Create a new class called "Message" in common for message handling with attributes
+  message_type, filename, function_name, line_number, timestamp.
+  (Logging will take line number and function name from where it is called, Here
+  the problem is we are logging the message from common so it takes a function name
+  and line number from common only. So it makes debugging complex. To avoid this
+  we have to find the function name, line number, file name and append this data
+  with message manually)
+```
+    from inspect import getframeinfo, stack
+    caller = getframeinfo(stack()[1][0])
+    caller.filename, caller.lineno, caller.function
+    (These are built in packages in python)
+```
+* Create a new class called "Log" in common with attributes log_level, message,
+  log_trace_back. Inherit Log class from Message class.
+* Create a new class called "Alert" in common with attributes alert_id, node_id,
+  resource, current_value, tags, type, severity, significance, ackedby, acked.
+  Inherit alert class from Message class. (loglevel for alert is always Warning, no need
+  any log information for alert because everything is predefined, so no need any log class help
+  for alert).
+* Create a new class called "FlightOperation" in common with attributes
+  flight_operation_id, message. Inherit FlightOperation class from Message class.
+* Using name retrieve the particular logging handler. Module specific logging
+  configuration should be used to log the messages, So we have to pass logging
+  handler name of the perticular module to Message Handler. Message handler will
+
+* Create a new function for Message class and process the object values.
+```
+  1. If the message type is Alert then store the message in log file with uuid (log level
+     for alert is always warning). and put the object in node_agent socket.
+  2. If the message type is flight operation then log the message in file with message
+     uuid (uuid is used to trace the filght operation).
+  3. If the type is Log then process then store message in log file..
+```
+  * Expected output format for log, alert, flight operation in log file:
+```
+  Message: ""%(asctime)s - %(filename)s - %(function)s - %(lineno)s - %(levelname)s
+  - %(message)s"
+
+  (No need to specify any log format in configuration file. We can remove the format
+  line from configuration files)
+```
+* Expected output format for the alert in node_agent socket:
+```
+  {
+     "timestamp"      : "time"
+     "alert_id"       : "string",
+     "node_id"        : "string",
+     "resource"       : "string",
+     "current-value"  : "string",
+     "significance"   : "string"
+     "severity"       : "string",
+     "type"           : "string",
+     "tags"           : "string",
+     "ackedby"        : "string",
+     "acked"          : "boolean"
```

Comment:
@GowthamShanmugam @shtripat comment looks like is required but I missed it somehow in my spec that actually talks about alert structure. But its my opinion that, instead of having changes to the alert structure and details around it be effected in 2 places this spec and also https://github.com/Tendrl/specifications/blob/master/specs/pluggable_alert_delivery.adoc. So I feel we can open a bug to capture this point in pluggable_alert_delivery.adoc and just refer that spec in this spec for details about alerts. @shtripat what do you suggest